### ISSUE TITLE
divide by zero encountered in divide in FasterRCNN

### ISSUE THREAD

**Issue by maxisme** [chainer/chainercv#578]
At the start of training the my FasterRCNN model I always get `divide by zero encountered in divide` errors although I am checking all my bbox's for validity in the dataset:

```
x1 = int(coords[0])
y1 = int(coords[1])
x2 = x1 + int(coords[2]) # x1 + width
y2 = y1 + int(coords[3]) # y1 + height

# check legal bbox
if x2 - x1 > 0 and y2 - y1 > 0:
    file_bboxs.append([x1,y1,x2,y2])
```

Errors:
```
/usr/local/lib/python2.7/dist-packages/chainercv/links/model/faster_rcnn/utils/bbox2loc.py:55: RuntimeWarning: divide by zero encountered in divide
  dy = (base_ctr_y - ctr_y) / height
/usr/local/lib/python2.7/dist-packages/chainercv/links/model/faster_rcnn/utils/bbox2loc.py:56: RuntimeWarning: divide by zero encountered in divide
  dx = (base_ctr_x - ctr_x) / width
/usr/local/lib/python2.7/dist-packages/chainercv/links/model/faster_rcnn/utils/bbox2loc.py:57: RuntimeWarning: divide by zero encountered in divide
  dh = xp.log(base_height / height)
/usr/local/lib/python2.7/dist-packages/chainercv/links/model/faster_rcnn/utils/bbox2loc.py:58: RuntimeWarning: divide by zero encountered in divide
  dw = xp.log(base_width / width)
```
I believe these errors are actually from the predicting of areas rather than my bbox?

**Comment by maxisme** [comment]
I am also encountering this part way through training:
```
/usr/local/lib/python2.7/dist-packages/chainercv/links/model/faster_rcnn/utils/proposal_creator.py:126: RuntimeWarning: invalid value encountered in greater_equal
  keep = np.where((hs >= min_size) & (ws >= min_size))[0]
````

**Comment by maxisme** [comment]
here is the main body of my code:
```
iterations = 50000
step_size = 30000
  
# initialise RCNN model with pretrained imagenet model
rcnn = FasterRCNNVGG16(n_fg_class=1, pretrained_model='imagenet')
rcnn.use_preset('evaluate')

model = FasterRCNNTrainChain(rcnn)
model.to_gpu(0)
chainer.cuda.get_device(0).use()

# initialise training data
train_data = TransformDataset(train_data, Transform(rcnn))
train_iter = chainer.iterators.SerialIterator(train_data, batch_size=1)

# initialise testing data
test_iter = chainer.iterators.SerialIterator(test_data, batch_size=1, repeat=False, shuffle=False)

# set model training paramas
optimizer = chainer.optimizers.MomentumSGD(lr=1e-3, momentum=0.9)
optimizer.setup(model)
optimizer.add_hook(chainer.optimizer.WeightDecay(rate=0.0005))

updater = chainer.training.updater.StandardUpdater(train_iter, optimizer, device=0)
trainer = training.Trainer(updater, (iterations, 'iteration'), out=config.localisation_model+"_logs")
trainer.extend(extensions.ExponentialShift('lr', 0.1), trigger=(step_size, 'iteration'))

## LOGGING ##
log_inter = (20, 'iteration')
trainer.extend(extensions.LogReport(trigger=log_inter))
trainer.extend(extensions.PrintReport(
    ['iteration', 'epoch', 'elapsed_time',
     'lr',
     'main/loss',
     'main/roi_loc_loss',
     'main/roi_cls_loss',
     'main/rpn_loc_loss',
     'main/rpn_cls_loss',
     'validation/main/map',
     ]), trigger=log_inter)
trainer.extend(extensions.ProgressBar(update_interval=10))
##############

# evaluates model by PASCAL VOC metric
trainer.extend(
    DetectionVOCEvaluator(
        test_iter, model.faster_rcnn, use_07_metric=True),
    trigger=ManualScheduleTrigger(
        [step_size, iterations], 'iteration'))

trainer.run()

serializers.save_npz(config.localisation_model, model)
```

**Comment by corochann** [comment]
Are you using your own dataset?
How large is the image size of your dataset?

I thought one reason that bbox's `height` or `width` become 0 is when the original image size it too big and when it is rescaled to be smaller, bbox becomes much smaller and becomes 0 `height` or `width`.

**Comment by maxisme** [comment]
I am using the WIDER dataset and for every single image before I store it I first find its `width` and `height` and then for the `bbox`s I perform the check :

```
                        if x1 >= 0 and y1 >= 0 and x1 <= width and y1 <= height and x2 >= 0 and y2 >= 0 and x2 <= width and y2 <= height and x2 - x1 > 0 and y2 - y1 > 0:
                            file_bboxs.append([x1,y1,x2,y2])
                        else:
                            print "ilegal bbox in "+ current_path
```
which excludes quite a lot. I realised I had not added the seed to numpy (`np.random.seed(347682)`) but unfortunately that made no difference when adding.

I have done what people recommend [here](https://github.com/smallcorgi/Faster-RCNN_TF/issues/158) and reduced the `lr`:
`optimizer = chainer.optimizers.MomentumSGD(lr=1e-5, momentum=0.9)`

literally nothing is working

**Comment by corochann** [comment]
Bbox is resized after you return bbox in your dataset,

For example at `Transform`: https://github.com/chainer/chainercv/blob/master/examples/faster_rcnn/train.py#L32

So just checking at the place of `DatasetMixin` class does not guarantee that bbox width or height is not 0 for proposal_creator's input.

**Comment by maxisme** [comment]
I have installed like this:

```
git clone https://github.com/chainer/chainercv.git
pip install -e chainercv
```

I got none of the above errors and the model is currently training:
```
iteration   epoch       elapsed_time  lr          main/loss   main/roi_loc_loss  main/roi_cls_loss  main/rpn_loc_loss  main/rpn_cls_loss  validation/main/map
20          0           10.7943                   3.63812     0.0772393          0.598424           2.1897             0.77276            
```
before `main/loss`, `main/roi_loc_loss`  and `main/roi_cls_loss` all returned `NaN`!! I think this may work 😀!! will let you know in `2:32:22.870733` 😉

**Comment by corochann** [comment]
Wow, good news.
So there may be a bug in old version and it was fixed in latest master branch.

**Comment by maxisme** [comment]
It worked!! Also the Transform function scales the bbox and the image so if it works on the original it will work on a scaled version too.

**Comment by maxisme** [comment]
@corochann thank you very much for your help btw!
